RFC.md

```diff
@@ -0,0 +1,413 @@
+	- Implement legacy checkpoint conversion to support old checkpoints
+
+### TransformerDecoder
+TransformerDecoder only highlighting the changes, current module [here](https://github.com/pytorch/torchtune/blob/0057fe7cf83e14f0b62538a8d4d20719e0a88639/torchtune/modules/transformer.py#L99). Discussion in the comments.
+
+```python
+class TransformerDecoder(nn.Module):
+    def __init__(
+        self,
+        *,
+        ...
+        # as opposed to layer and num_layers
+        layers: List[nn.Module],
+        ...
+        # output is generalized and hidden_idx is added to address #1017
+        output: Callable
+        hidden_idx: Optional[List[int]] = None
+    ) -> None:
+        super().__init__()
+        # We could keep an optional layer + num_layers args for BC
+        # Otherwise this change requires builders to go from
+        # TransformerDecoder(layer, 32)
+        # To
+        # for _ in range(32):
+	    #    layers.append(layer)
+	    # TransformerDecoder(layers)
+        ...
+
+    def forward(
+        self,
+        tokens: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> [Tensor, List[Tensor]]:
+        ...
+
+		# Move max seq length check from attention to here
+        if seq_len > self.max_seq_len:
+            raise ValueError("Too long")
+
+        ...
+
+		hidden = []
+        for i, layer in enumerate(self.layers):
+	        if i in self.hidden_idx:
+		        hidden.append(h)
+            h = layer(
+                h,
+                mask=mask,
+                encoder_input=encoder_input,
+                encoder_mask=encoder_mask,
+                input_pos=input_pos,
+            )
+
+        # shape: [b, s, d]
+        h = self.norm(h)
+
+        output = self.output(h).float()
+        # this unifies the output with ViT but breaks recipes
+        # Alternative:
+        # output = output if not hidden else hidden + [output]
+        return output, hidden
+```
+@SalmanMohammadi does this design address all your needs from #1017
+### TransformerSelfAttentionLayer
+TransformerDecoderLayer only highlighting the changes, current module [here](https://github.com/pytorch/torchtune/blob/0057fe7cf83e14f0b62538a8d4d20719e0a88639/torchtune/modules/transformer.py#L15). Discussion in the comments.
+
+```python
+class TransformerSelfAttentionLayer(nn.Module):
+    def __init__(
+        self,
+        attn: GroupedQueryAttention, # Updated Attn module
+        mlp: nn.Module,
+        *,
+        # sa_norm renamed since it's no longer only self attention
+        # this requires repo wide changes for sa_norm -> attn_norm
+        attn_norm: Optional[nn.Module] = None,
```

Comment:
can this still remain `sa_norm` since we're still in TransformerSelfAttentionLayer? I get that you'd want to keep the inits uniform across SA and XA layers, but if we are designing to accept any type of attention layer, would it be reasonable to relax this constraint? And then you won't have any checkpoint / BC breaking changes.

Reply:
I guess it could remain as sa_norm. I prefer having these be the same for both layer types but there's nothing enforcing that.

Reply:
Personally I would be inclined to break BC here for the sake of consistency across self- and cross-attention, otherwise it will just introduce extra remapping across the repo moving forward that will be more of a headache (especially since we are already breaking BC with these changes, we might as well do this while we can get away with it).